<a href="https://colab.research.google.com/github/zmostafa/CarND-Keras-Lab/blob/master/traffic-sign-classification-with-keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Traffic Sign Classification with Keras

Keras exists to make coding deep neural networks simpler. To demonstrate just how easy it is, you’re going to use Keras to build a convolutional neural network in a few dozen lines of code.

You’ll be connecting the concepts from the previous lessons to the methods that Keras provides.

## Dataset

The network you'll build with Keras is similar to the example in Keras’s GitHub repository that builds out a [convolutional neural network for MNIST](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py). 

However, instead of using the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset, you're going to use the [German Traffic Sign Recognition Benchmark](http://benchmark.ini.rub.de/?section=gtsrb&subsection=news) dataset that you've used previously.

You can download pickle files with sanitized traffic sign data here:

In [1]:
from urllib.request import urlretrieve
from os.path import isfile
from tqdm import tqdm

class DLProgress(tqdm):
    last_block = 0

    def hook(self, block_num=1, block_size=1, total_size=None):
        self.total = total_size
        self.update((block_num - self.last_block) * block_size)
        self.last_block = block_num

if not isfile('train.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Train Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/train.p',
            'train.p',
            pbar.hook)

if not isfile('test.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Test Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/test.p',
            'test.p',
            pbar.hook)

print('\nTraining and Test data downloaded.')

Train Dataset: 120MB [00:03, 33.4MB/s]                           
Test Dataset: 38.8MB [00:01, 23.8MB/s]                           


Training and Test data downloaded.


## Overview

Here are the steps you'll take to build the network:

1. Load the training data.
2. Preprocess the data.
3. Build a feedforward neural network to classify traffic signs.
4. Build a convolutional neural network to classify traffic signs.
5. Evaluate the final neural network on testing data.

Keep an eye on the network’s accuracy over time. Once the accuracy reaches the 98% range, you can be confident that you’ve built and trained an effective model.

In [2]:
import pickle
import numpy as np
import math

# Fix error with TF and Keras
import tensorflow as tf
tf.python.control_flow_ops = tf

print('Modules loaded.')

Modules loaded.


## Load the Data

Start by importing the data from the pickle file.

In [0]:
with open('train.p', 'rb') as f:
    data = pickle.load(f)

# TODO: Load the feature data to the variable X_train
    X_train = data['features']
# TODO: Load the label data to the variable y_train
    y_train = data['labels']

In [4]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert np.array_equal(X_train, data['features']), 'X_train not set to data[\'features\'].'
assert np.array_equal(y_train, data['labels']), 'y_train not set to data[\'labels\'].'
print('Tests passed.')

Tests passed.


## Preprocess the Data

1. Shuffle the data
2. Normalize the features using Min-Max scaling between -0.5 and 0.5
3. One-Hot Encode the labels

### Shuffle the data
Hint: You can use the [scikit-learn shuffle](http://scikit-learn.org/stable/modules/generated/sklearn.utils.shuffle.html) function to shuffle the data.

In [0]:
from sklearn.utils import shuffle
# TODO: Shuffle the data
X_train , y_train = shuffle(X_train,y_train)

In [6]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert X_train.shape == data['features'].shape, 'X_train has changed shape. The shape shouldn\'t change when shuffling.'
assert y_train.shape == data['labels'].shape, 'y_train has changed shape. The shape shouldn\'t change when shuffling.'
assert not np.array_equal(X_train, data['features']), 'X_train not shuffled.'
assert not np.array_equal(y_train, data['labels']), 'y_train not shuffled.'
print('Tests passed.')

Tests passed.


### Normalize the features
Hint: You solved this in [TensorFlow lab](https://github.com/udacity/CarND-TensorFlow-Lab/blob/master/lab.ipynb) Problem 1.

In [0]:
# TODO: Normalize the data features to the variable X_normalized
def normalize(img):
  a = -0.5
  b = 0.5
  return (a + (((img - 0) * (b - a)) / (255  - 0)))
  # return (img - 128.) / 128.

X_normalized = np.array([normalize(img) for img in X_train])

In [8]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert math.isclose(np.min(X_normalized), -0.5, abs_tol=1e-5) and math.isclose(np.max(X_normalized), 0.5, abs_tol=1e-5), 'The range of the training data is: {} to {}.  It must be -0.5 to 0.5'.format(np.min(X_normalized), np.max(X_normalized))
print('Tests passed.')

Tests passed.


### One-Hot Encode the labels
Hint: You can use the [scikit-learn LabelBinarizer](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html) function to one-hot encode the labels.

In [0]:
# TODO: One Hot encode the labels to the variable y_one_hot
from sklearn.preprocessing import LabelBinarizer
label_binarizer = LabelBinarizer()
y_one_hot = label_binarizer.fit_transform(y_train)

In [10]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
import collections

assert y_one_hot.shape == (39209, 43), 'y_one_hot is not the correct shape.  It\'s {}, it should be (39209, 43)'.format(y_one_hot.shape)
assert next((False for y in y_one_hot if collections.Counter(y) != {0: 42, 1: 1}), True), 'y_one_hot not one-hot encoded.'
print('Tests passed.')

Tests passed.


## Keras Sequential Model
```python
from keras.models import Sequential

# Create the Sequential model
model = Sequential()
```
The `keras.models.Sequential` class is a wrapper for the neural network model. Just like many of the class models in scikit-learn, it provides common functions like `fit()`, `evaluate()`, and `compile()`.  We'll cover these functions as we get to them.  Let's start looking at the layers of the model.

## Keras Layer
A Keras layer is just like a neural network layer.  It can be fully connected, max pool, activation, etc.  You can add a layer to the model using the model's `add()` function.  For example, a simple model would look like this:
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten

# Create the Sequential model
model = Sequential()

# 1st Layer - Add a flatten layer
model.add(Flatten(input_shape=(32, 32, 3)))

# 2nd Layer - Add a fully connected layer
model.add(Dense(100))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer
model.add(Dense(60))

# 5th Layer - Add a ReLU activation layer
model.add(Activation('relu'))
```
Keras will automatically infer the shape of all layers after the first layer.  This means you only have to set the input dimensions for the first layer.

The first layer from above, `model.add(Flatten(input_shape=(32, 32, 3)))`, sets the input dimension to (32, 32, 3) and output dimension to (3072=32\*32\*3).  The second layer takes in the output of the first layer and sets the output dimenions to (100).  This chain of passing output to the next layer continues until the last layer, which is the output of the model.

## Build a Multi-Layer Feedforward Network

Build a multi-layer feedforward neural network to classify the traffic sign images.

1. Set the first layer to a `Flatten` layer with the `input_shape` set to (32, 32, 3)
2. Set the second layer to `Dense` layer width to 128 output. 
3. Use a ReLU activation function after the second layer.
4. Set the output layer width to 43, since there are 43 classes in the dataset.
5. Use a softmax activation function after the output layer.

To get started, review the Keras documentation about [models](https://keras.io/models/sequential/) and [layers](https://keras.io/layers/core/).

The Keras example of a [Multi-Layer Perceptron](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py) network is similar to what you need to do here. Use that as a guide, but keep in mind that there are a number of differences.

In [11]:
from keras.models import Sequential
from keras.layers.core import Dense, Flatten , Activation
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
# 1st Layer - Add a flatten layer
model.add(Flatten(input_shape=(32,32,3)))

# 2nd Layer - Add a fully connected layer with output = 128
model.add(Dense(128))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer with output = 43
model.add(Dense(43))

# 5th Layer - Add a softmax Activiation layer
model.add(Activation('softmax'))

Using TensorFlow backend.


In [12]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax

def check_layers(layers, true_layers):
    assert len(true_layers) != 0, 'No layers found'
    for layer_i in range(len(layers)):
        assert isinstance(true_layers[layer_i], layers[layer_i]), 'Layer {} is not a {} layer'.format(layer_i+1, layers[layer_i].__name__)
    assert len(true_layers) == len(layers), '{} layers found, should be {} layers'.format(len(true_layers), len(layers))

check_layers([Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[1].output_shape == (None, 128), 'Second layer output is wrong, it should be (128)'
assert model.layers[2].activation == relu, 'Third layer not a relu activation layer'
assert model.layers[3].output_shape == (None, 43), 'Fourth layer output is wrong, it should be (43)'
assert model.layers[4].activation == softmax, 'Fifth layer not a softmax activation layer'
print('Tests passed.')

Tests passed.


## Training a Sequential Model
You built a multi-layer neural network in Keras, now let's look at training a neural network.
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation

model = Sequential()
...

# Configures the learning process and metrics
model.compile('sgd', 'mean_squared_error', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(X_train_data, Y_train_data, batch_size=128, epochs=2, validation_split=0.2, verbose=2)

# Calculate test score
test_score = model.evaluate(X_test_data, Y_test_data)
```
The code above configures, trains, and tests the model.  The line `model.compile('sgd', 'mean_squared_error', ['accuracy'])` configures the model's optimizer to `'sgd'`(stochastic gradient descent), the loss to `'mean_squared_error'`, and the metric to `'accuracy'`.  

You can find more optimizers [here](https://keras.io/optimizers/), loss functions [here](https://keras.io/objectives/#available-objectives), and more metrics [here](https://keras.io/metrics/#available-metrics).

To train the model, use the `fit()` function as shown in `model.fit(X_train_data, Y_train_data, batch_size=128, epochs=2, validation_split=0.2, verbose=2)`.  The `validation_split` parameter will split a percentage of the training dataset to be used to validate the model.  Typically you won't have to change the `verbose` parameter but in Jupyter notebooks the update animation can crash the notebook so we set `verbose=2`, this limits the animation to only update after an epoch is complete. The model can be further tested with the test dataset using the `evaluation()` function as shown in the last line.

## Train the Network

1. Compile the network using adam optimizer and categorical_crossentropy loss function.
2. Train the network for ten epochs and validate with 20% of the training data.

In [13]:
# TODO: Compile and train the model here.
model.compile('adam','categorical_crossentropy',['acc'])
history = model.fit(X_normalized, y_one_hot, epochs=10, validation_split=0.2, verbose=2)
 

Train on 31367 samples, validate on 7842 samples
Epoch 1/10
 - 7s - loss: 1.4919 - acc: 0.5981 - val_loss: 0.9241 - val_acc: 0.7353
Epoch 2/10
 - 7s - loss: 0.6711 - acc: 0.8114 - val_loss: 0.5323 - val_acc: 0.8490
Epoch 3/10
 - 7s - loss: 0.4992 - acc: 0.8581 - val_loss: 0.5238 - val_acc: 0.8521
Epoch 4/10
 - 7s - loss: 0.4264 - acc: 0.8771 - val_loss: 0.4340 - val_acc: 0.8791
Epoch 5/10
 - 7s - loss: 0.3579 - acc: 0.8968 - val_loss: 0.5487 - val_acc: 0.8417
Epoch 6/10
 - 7s - loss: 0.3278 - acc: 0.9047 - val_loss: 0.2918 - val_acc: 0.9193
Epoch 7/10
 - 7s - loss: 0.3053 - acc: 0.9097 - val_loss: 0.3221 - val_acc: 0.9061
Epoch 8/10
 - 7s - loss: 0.2749 - acc: 0.9199 - val_loss: 0.3310 - val_acc: 0.9077
Epoch 9/10
 - 7s - loss: 0.2559 - acc: 0.9262 - val_loss: 0.2662 - val_acc: 0.9251
Epoch 10/10
 - 7s - loss: 0.2431 - acc: 0.9300 - val_loss: 0.5148 - val_acc: 0.8573


In [14]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.optimizers import Adam

assert model.loss == 'categorical_crossentropy', 'Not using categorical_crossentropy loss function'
assert isinstance(model.optimizer, Adam), 'Not using adam optimizer'
assert len(history.history['acc']) == 10, 'You\'re using {} epochs when you need to use 10 epochs.'.format(len(history.history['acc']))

assert history.history['acc'][-1] > 0.92, 'The training accuracy was: %.3f. It shoud be greater than 0.92' % history.history['acc'][-1]
assert history.history['val_acc'][-1] > 0.85, 'The validation accuracy is: %.3f. It shoud be greater than 0.85' % history.history['val_acc'][-1]
print('Tests passed.')

Tests passed.


## Convolutions
1. Re-construct the previous network
2. Add a [convolutional layer](https://keras.io/layers/convolutional/#convolution2d) with 32 filters, a 3x3 kernel, and valid padding before the flatten layer.
3. Add a ReLU activation after the convolutional layer.

Hint 1: The Keras example of a [convolutional neural network](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) for MNIST would be a good example to review.

In [0]:
# TODO: Re-construct the network and add a convolutional layer before the flatten layer.
from keras.models import Sequential
from keras.layers.core import Dense, Flatten , Activation
from keras.layers.convolutional import Conv2D, MaxPooling2D
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
# 1st Layer - Add a flatten layer
model.add(Conv2D(32,(3,3),padding='valid',input_shape=(32,32,3)))
model.add(Activation('relu'))
model.add(Flatten())

# 2nd Layer - Add a fully connected layer with output = 128
model.add(Dense(128))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer with output = 43
model.add(Dense(43))

# 5th Layer - Add a softmax Activiation layer
model.add(Activation('softmax'))

In [23]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Conv2D

check_layers([Conv2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[0].filters == 32, 'Wrong number of filters, it should be 32'
assert model.layers[0].kernel_size[0] == model.layers[0].kernel_size[1] == 3, 'Kernel size is wrong, it should be a 3x3'
assert model.layers[0].padding == 'valid', 'Wrong padding, it should be valid'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, epochs=2, validation_split=0.2, verbose=2)
assert(history.history['val_accuracy'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_accuracyc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
 - 36s - loss: 0.0807 - accuracy: 0.9804 - val_loss: 0.1569 - val_accuracy: 0.9587
Epoch 2/2
 - 39s - loss: 0.0602 - accuracy: 0.9844 - val_loss: 0.1392 - val_accuracy: 0.9620
Tests passed.


## Pooling
1. Re-construct the network
2. Add a 2x2 [max pooling layer](https://keras.io/layers/pooling/#maxpooling2d) immediately following your convolutional layer.

In [0]:
# TODO: Re-construct the network and add a pooling layer after the convolutional layer.
from keras.models import Sequential
from keras.layers.core import Dense, Flatten , Activation
from keras.layers.convolutional import Conv2D, MaxPooling2D
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
# 1st Layer - Add a Conv with 32 filters and input 32x32x3
model.add(Conv2D(32,(3,3),padding='valid',input_shape=(32,32,3)))

# 2nd Layer - Add a MaxPooling layer with 2x2 kernel size
model.add(MaxPooling2D((2,2)))

# 3rd Layer - Add a activation layer
model.add(Activation('relu'))

# 4th Layer - Flatten output from previous layer
model.add(Flatten())

# 5th Layer - Add a fully connected layer with output = 128
model.add(Dense(128))

# 6th Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 7th Layer - Add a fully connected layer with output = 43
model.add(Dense(43))

# 8th Layer - Add a softmax Activiation layer
model.add(Activation('softmax'))

In [28]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Conv2D
from keras.layers.pooling import MaxPooling2D

check_layers([Conv2D, MaxPooling2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[1].pool_size == (2, 2), 'Second layer must be a max pool layer with pool size of 2x2'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, epochs=2, validation_split=0.2, verbose=2)
assert(history.history['val_accuracy'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_accuracy'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
 - 21s - loss: 0.2519 - accuracy: 0.9407 - val_loss: 0.2490 - val_accuracy: 0.9384
Epoch 2/2
 - 21s - loss: 0.1579 - accuracy: 0.9633 - val_loss: 0.1668 - val_accuracy: 0.9625
Tests passed.


## Dropout
1. Re-construct the network
2. Add a [dropout](https://keras.io/layers/core/#dropout) layer after the pooling layer. Set the dropout rate to 50%.

In [0]:
# TODO: Re-construct the network and add dropout after the pooling layer.
from keras.models import Sequential
from keras.layers.core import Dense, Flatten , Activation, Dropout
from keras.layers.convolutional import Conv2D, MaxPooling2D
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
# 1st Layer - Add a Conv with 32 filters and input 32x32x3
model.add(Conv2D(32,(3,3),padding='valid',input_shape=(32,32,3)))

# 2nd Layer - Add a MaxPooling layer with 2x2 kernel size
model.add(MaxPooling2D((2,2)))

# 3rd Layer - Add a Dropout
model.add(Dropout(0.5))

# 4th Layer - Add a activation layer
model.add(Activation('relu'))

# 5th Layer - Flatten output from previous layer
model.add(Flatten())

# 6th Layer - Add a fully connected layer with output = 128
model.add(Dense(128))

# 7th Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 8th Layer - Add a fully connected layer with output = 43
model.add(Dense(43))

# 9th Layer - Add a softmax Activiation layer
model.add(Activation('softmax'))

In [31]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.layers.convolutional import Conv2D
from keras.layers.pooling import MaxPooling2D

check_layers([Conv2D, MaxPooling2D, Dropout, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[2].rate == 0.5, 'Third layer should be a Dropout of 50%'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, epochs=2, validation_split=0.2, verbose=2)
assert(history.history['val_accuracy'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_accuracy'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
 - 23s - loss: 1.6054 - accuracy: 0.5775 - val_loss: 0.7446 - val_accuracy: 0.7913
Epoch 2/2
 - 23s - loss: 0.5200 - accuracy: 0.8651 - val_loss: 0.3560 - val_accuracy: 0.9129
Tests passed.


## Optimization
Congratulations! You've built a neural network with convolutions, pooling, dropout, and fully-connected layers, all in just a few lines of code.

Have fun with the model and see how well you can do! Add more layers, or regularization, or different padding, or batches, or more training epochs.

What is the best validation accuracy you can achieve?

In [33]:
# TODO: Build a model
from keras.models import Sequential
from keras.layers.core import Dense, Flatten , Activation, Dropout
from keras.layers.convolutional import Conv2D, MaxPooling2D
model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
# 1st Layer - Add a Conv with 32 filters and input 32x32x3
model.add(Conv2D(32,(3,3),padding='valid',activation='relu',input_shape=(32,32,3)))

# 2nd Layer - Add a MaxPooling layer with 2x2 kernel size
model.add(MaxPooling2D((2,2)))

# 3rd Layer - Add a Dropout
model.add(Dropout(0.5))

# 4th Layer - Add a activation layer
model.add(Activation('relu'))

model.add(Conv2D(16,(3,3),padding='valid',activation='relu'))
model.add(MaxPooling2D((2,2)))

model.add(Dropout(0.25))
# 5th Layer - Flatten output from previous layer
model.add(Flatten())

# 6th Layer - Add a fully connected layer with output = 128
model.add(Dense(128))
model.add(Dropout(0.5))
# 7th Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 8th Layer - Add a fully connected layer with output = 43
model.add(Dense(43))

# 9th Layer - Add a softmax Activiation layer
model.add(Activation('softmax'))
# TODO: Compile and train the model

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, epochs=5, validation_split=0.2, verbose=2)

Train on 31367 samples, validate on 7842 samples
Epoch 1/5
 - 28s - loss: 2.8320 - accuracy: 0.2338 - val_loss: 1.7709 - val_accuracy: 0.5194
Epoch 2/5
 - 28s - loss: 1.5468 - accuracy: 0.5245 - val_loss: 0.8559 - val_accuracy: 0.7911
Epoch 3/5
 - 28s - loss: 1.0365 - accuracy: 0.6752 - val_loss: 0.5269 - val_accuracy: 0.8910
Epoch 4/5
 - 28s - loss: 0.8040 - accuracy: 0.7423 - val_loss: 0.3702 - val_accuracy: 0.9269
Epoch 5/5
 - 28s - loss: 0.6744 - accuracy: 0.7848 - val_loss: 0.3227 - val_accuracy: 0.9430


**Best Validation Accuracy:** (fill in here)

## Testing
Once you've picked out your best model, it's time to test it.

Load up the test data and use the [`evaluate()` method](https://keras.io/models/model/#evaluate) to see how well it does.

Hint 1: The `evaluate()` method should return an array of numbers. Use the [`metrics_names`](https://keras.io/models/model/) property to get the labels.

In [36]:
# TODO: Load test data
with open('test.p', 'rb') as f:
    data = pickle.load(f)

# TODO: Load the feature data to the variable X_train
    X_test = data['features']
# TODO: Load the label data to the variable y_train
    y_test = data['labels']

# TODO: Preprocess data & one-hot encode the labels
X_test_normalized = np.array([normalize(img) for img in X_test])
y_test_hot = label_binarizer.fit_transform(y_test)
# TODO: Evaluate model on test data
metrics = model.evaluate(X_test_normalized,y_test_hot)
for metric_i in range(len(model.metrics_names)):
    metric_name = model.metrics_names[metric_i]
    metric_value = metrics[metric_i]
    print('{}: {}'.format(metric_name, metric_value))

12630/12630 [==============================] - 3s 277us/step
loss: 0.4737065553759547
accuracy: 0.8907363414764404


**Test Accuracy:** 

loss: 0.4737065553759547

accuracy: 0.8907363414764404

## Summary
Keras is a great tool to use if you want to quickly build a neural network and evaluate performance.